src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -348,57 +356,52 @@ impl<'a, 'tcx, 'rcx> DocFolder for LinkCollector<'a, 'tcx, 'rcx> {
+                        }
+
+                        if let Ok(type_def) =
                             self.resolve(path_str, false, &current_item, parent_node)
                         {
-                            // It is imperative we search for not-a-value first
-                            // Otherwise we will find struct ctors for when we are looking
-                            // for structs, and the link won't work if there is something in
-                            // both namespaces.
-                            if let Ok(value_def) =
-                                self.resolve(path_str, true, &current_item, parent_node)
-                            {
-                                let kind = value_ns_kind(value_def.0, path_str);
-                                if let Some((value_kind, value_disambig)) = kind {
-                                    let (type_kind, article, type_disambig)
-                                        = type_ns_kind(type_def.0, path_str);
-                                    ambiguity_error(cx, &item.attrs, path_str,
-                                                    article, type_kind, &type_disambig,
-                                                    "a", value_kind, &value_disambig);
-                                    continue;
-                                }
-                            }
-                            type_def
-                        } else if let Ok(value_def) =
+                            candidates.push((type_def, Namespace::Type));
+                        }
+
+                        if let Ok(value_def) =
                             self.resolve(path_str, true, &current_item, parent_node)
                         {
-                            value_def
-                        } else {
+                            // Structs, variants, and mods exist in both namespaces, skip them.
+                            match value_def.0 {
+                                Def::StructCtor(..)
+                                | Def::Mod(..)
+                                | Def::Variant(..)
+                                | Def::VariantCtor(..)
+                                | Def::SelfCtor(..) => (),
+                                _ => candidates.push((value_def, Namespace::Value)),
+                            }
+                        }
+
+                        if candidates.len() == 1 {
+                            candidates.remove(0).0
+                        } else if candidates.is_empty() {
                             resolution_failure(cx, &item.attrs, path_str, &dox, link_range);
                             // this could just be a normal link
                             continue;
+                        } else {
+                            let candidates = candidates.into_iter().map(|((def, _), ns)| {
+                                (def, ns)
+                            }).collect::<Vec<_>>();
```

Comment:
Looks like the `candidates` vector can be collected from `(Def, Namespace)` pairs right away and don't need to be re-collected.

Reply:
Wait, there's a better data structure for this than `Vec<(Def, Namespace)>`, it's called `struct PerNS`.

Reply:
`PerNS` doesn't seem that useful here since there's no way to iterate the values *and* the namespaces. We have to know what namespace the value came from for error reporting.